### PR TITLE
[AIRFLOW-1474] Add dag_id regex feature for `airflow clear` command

### PR DESCRIPTION
**Pull request by jgao54**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [AIRFLOW-1474](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1474

### Description
- [ ] Here are some details about my PR, including screenshots of any UI changes:
The 'airflow clear' CLI command is currently limited to clearing a single DAG per operation. It would be useful to add the capability to clear multiple DAGs per operation using regex, similar to how task_id can be filtered via regex.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
    - ClearTasksTest.test_dags_clear
    - CliTests.test_get_dags

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: airflow/models.py [hunk 1]
```diff
@@ -3218,11 +3218,10 @@ def clear(
             # Crafting the right filter for dag_id and task_ids combo
             conditions = []
             for dag in self.subdags + [self]:
-                if dag.task_ids:
```
Comment (jgao54) [inline airflow/models.py:3221]: This line is removed to fix a bug where when `-t TASK_REGEX` flag is specified, if the task_regex does not match any task, then dag.task_ids = [] results in conditions = [], and consequently no filters are applied.

**Comment by criccomini** [comment]
@bolkedebruin would like your feedback here as well

**Comment by criccomini** [comment]
Use case for this, btw, is sometimes we have an outage with a downstream system (e.g. Google Cloud Storage/BigQuery), and we get 100s of DAG failures in the span of a few minutes. We eventually halt things, but then we need to go clear everything out. Having a regex would make this much easier.

**Comment by criccomini** [comment]
LGTM, provided tests pass, but want to make sure @bolkedebruin is aware of change to CLI

File: tests/core.py [hunk 2]
```diff
@@ -1354,6 +1354,19 @@ def test_subdag_clear(self):
...
+        self.assertEqual(len(dags), 1)
+
+        dags = cli.get_dags(self.parser.parse_args(['clear', 'subdag', '-dx', '-c']))
+        self.assertGreater(len(dags), 1)
+
+        try:
+            cli.get_dags(self.parser.parse_args(['clear', 'foobar', '-dx', '-c']))
+            self.fail("dag with regex 'foobar' should not exist!")
+        except AirflowException:
+            pass # expected exception
```
Comment (skudriashev) [inline tests/core.py]: Why not to use `assertRaises`?

Comment (jgao54) [reply tests/core.py]: Fixed, thanks! :)

**Comment by criccomini** [comment]
Seems tests failed. :(
```
1) ERROR: test_dags_clear (tests.ClearTasksTest)
----------------------------------------------------------------------
   Traceback (most recent call last):
    tests/models.py line 1313 in test_dags_clear
      tis[failed_dag_idx].state = State.FAILED
   IndexError: list index out of range
```

**Comment by jgao54** [comment]
@criccomini  https://twitter.com/codinghorror/status/506010907021828096, hehe 😅

File: airflow/models.py [hunk 3]
```diff
@@ -3267,6 +3265,61 @@ def clear(
...
+        if dry_run:
+            return all_tis
+
+        count = len(all_tis)
+        do_it = True
+        if count == 0:
+            print("Nothing to clear.")
+            return 0
+        if confirm_prompt:
+            ti_list = "\n".join([str(t) for t in all_tis])
```
Comment (criccomini) [inline airflow/models.py:3301]: Apparently, this isn't used, according to landscape?

**Review by criccomini — COMMENTED** [review]
I'm going to let this sit one more day for @bolkedebruin before I move ahead and merge.

@jgao54 could you fix some of the landscape issues? Some are not easily fixable, but some should be easy to knock out.

File: airflow/models.py (hunk 3 above)
Comment (jgao54) [reply airflow/models.py:3301]: landscape isn't sophisticated enough to deal with locals(), so it thinks it's not used:
```ti_list = "\n".join([str(t) for t in tis])
question = (
    "You are about to delete these {count} tasks:\n"
    "{ti_list}\n\n"
    "Are you sure? (yes/no): ").format(**locals())
```

But yeah I can do a work-around.

Comment (criccomini) [reply airflow/models.py:3301]: Ah, sigh. :P

**Comment by criccomini** [comment]
LGTM merged!

**Pull request closed without merging**